node/util_test.go

```diff
@@ -0,0 +1,123 @@
+package node
+
+import (
+	"context"
+	"errors"
+	"path/filepath"
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/stretchr/testify/require"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/config"
+	"github.com/spacemeshos/go-spacemesh/log"
+)
+
+// NewTestNetwork creates a network of fully connected nodes.
+func NewTestNetwork(t *testing.T, conf config.Config, l log.Log, size int) []*TestApp {
+	// We need to set this global state
+	types.SetLayersPerEpoch(conf.LayersPerEpoch)
+	types.SetNetworkHRP(conf.NetworkHRP) // set to generate coinbase
+
+	// To save an epoch of startup time, we bootstrap (meaning we manually set
+	// it) the beacon for epoch 2 so that in epoch 3 hare can start.
+	bootstrapEpoch := (types.GetEffectiveGenesis() + 1).GetEpoch()
+	bootstrapBeacon := types.Beacon{}
+	genesis := conf.Genesis.GenesisID()
+	copy(bootstrapBeacon[:], genesis[:])
+
+	ctx, cancel := context.WithTimeout(context.Background(), 20*time.Second)
+	g := errgroup.Group{}
+	var apps []*TestApp
+
+	t.Cleanup(func() {
+		cancel()
+		// Wait for nodes to shutdown
+		g.Wait()
+
+		ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
+		defer cancel()
+		for _, a := range apps {
+			a.Cleanup(ctx)
+		}
+	})
+
+	for i := 0; i < size; i++ {
+		// Copy config, services don't modify their config, so we just need to
+		// be careful here when we modify any pointer values in the config.
+		c := conf
+		dir := t.TempDir()
+		c.DataDirParent = dir
+		c.SMESHING.Opts.DataDir = dir
+		c.SMESHING.CoinbaseAccount = types.GenerateAddress([]byte(strconv.Itoa(i))).String()
+		c.FileLock = filepath.Join(c.DataDirParent, "LOCK")
+
+		app := NewApp(t, &c, l)
+		instanceIndex := i
+		g.Go(func() error {
+			err := app.Start(ctx)
+			if err != nil && !errors.Is(err, context.Canceled) {
+				t.Logf("failed to start instance %d: %v", instanceIndex, err)
+			}
+			return err
+		})
```

Comment:
Should the test be stopped as soon as the first node encounters some error? Maybe:

```go
g.Go(func() error {
	err := app.Start(ctx)
	if err != nil && !errors.Is(err, context.Canceled) {
		require.Fail(t, "node instantiation", "failed to start instance %d: %v", instanceIndex, err)
	}
	return nil
})
```

This would end the test (and shut down all nodes) immediately as soon as the first has some issue.

Reply:
Then it is slightly more involved: `errgroup.WithContext` can create a sub-context that is canceled when the first go routine returns an error. If you pass that context to `app.Start` all nodes should be shut down as soon as the first returns a non-nil error.

Reply:
Sounds good!